app/models/project.rb

```diff
@@ -513,30 +513,37 @@ def download_registry_users
     end
   end
 
-  def self.find_with_includes!(platform, name, includes)
-    # this clunkiness is because .includes() doesn't allow zero args and we want
-    # to allow that.
-    
+  def self.find_best(platform, name, includes=[])
+    begin
+      find_exact!(platform, name, includes)
+    rescue ActiveRecord::RecordNotFound
+      find_fuzzy!(platform, name, includes)
+    end
+  rescue ActiveRecord::RecordNotFound
+    nil
+  end
+
+  def self.find_exact!(platform, name, includes=[])
+    visible
+      .platform(platform)
+      .where(name: name)
+      .includes(includes.present? ? includes : nil)
+      .first!
+  end
+
+  def self.find_fuzzy!(platform, name, includes=[])
     platform_class = PackageManager::Base.find(platform)
     raise ActiveRecord::RecordNotFound if platform_class.nil?
     project_find_names = platform_class.project_find_names(name)
-    
-    query = self.visible.platform(platform).where(name: project_find_names)
-    query = query.includes(*includes) unless includes.empty?
-    project = query.first
-    if project.nil?
-      query = self.visible.lower_platform(platform).lower_name(name)
-      query = query.includes(*includes) unless includes.empty?
-      project = query.first
-    end
-    raise ActiveRecord::RecordNotFound if project.nil?
-    raise ActiveRecord::RecordNotFound if project.status == 'Hidden'
-    project
+
+    visible.lower_platform(platform).lower_name(name).first ||
+      visible.platform(platform).where(name: project_find_names).first!
```

Comment:
These seem backwards from the original logic. I think we should only attempt the lowercase query if we don't find a project with `project_find_names`. It's also a bit odd that platform is downcased sometimes but not always.

Reply:
I think I only want to do `project_find_names` if we don't find anything with the lowercase query - it's a potentially expensive operation (eg: making a network call for Go)